### ISSUE TITLE
GarbageCollection exception on 1.12.12 deleting files

### ISSUE THREAD

**Issue by chrisbardon** [rtyley/bfg-repo-cleaner#159]
I'm trying to delete files from a repo, and I've run through one pass, but on a second pass (after running prune/GC), I'm getting this:

java -jar bfg-1.12.12.jar --delete-files *.sdf ice

Using repo : C:\ice-subgit-6.16\ice.git

This repo has been processed by The BFG before! Will prune repo before proceedin
g - to avoid unnecessary cleaning work on unused objects...
Exception in thread "main" org.eclipse.jgit.api.errors.JGitInternalException: Ga
rbage collection failed.
        at org.eclipse.jgit.api.GarbageCollectCommand.call(GarbageCollectCommand
.java:192)
        at com.madgag.git.bfg.cli.Main$$anonfun$1.apply(Main.scala:49)
        at com.madgag.git.bfg.cli.Main$$anonfun$1.apply(Main.scala:34)
        at scala.Option.map(Option.scala:146)
        at com.madgag.git.bfg.cli.Main$.delayedEndpoint$com$madgag$git$bfg$cli$M
ain$1(Main.scala:33)
        at com.madgag.git.bfg.cli.Main$delayedInit$body.apply(Main.scala:27)
        at scala.Function0$class.apply$mcV$sp(Function0.scala:34)
        at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:
12)
        at scala.App$$anonfun$main$1.apply(App.scala:76)
        at scala.App$$anonfun$main$1.apply(App.scala:76)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at scala.collection.generic.TraversableForwarder$class.foreach(Traversab
leForwarder.scala:35)
        at scala.App$class.main(App.scala:76)
        at com.madgag.git.bfg.cli.Main$.main(Main.scala:27)
        at com.madgag.git.bfg.cli.Main.main(Main.scala)
Caused by: org.eclipse.jgit.errors.MissingObjectException: Missing unknown a0c09
5ae881da8c015beef79c03878a0f0cd7629
        at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor
.java:145)
        at org.eclipse.jgit.lib.ObjectReader$1.open(ObjectReader.java:299)
        at org.eclipse.jgit.revwalk.RevWalk$2.next(RevWalk.java:971)
        at org.eclipse.jgit.internal.storage.pack.PackWriter.findObjectsToPack(P
ackWriter.java:1660)
        at org.eclipse.jgit.internal.storage.pack.PackWriter.preparePack(PackWri
ter.java:756)
        at org.eclipse.jgit.internal.storage.pack.PackWriter.preparePack(PackWri
ter.java:717)
        at org.eclipse.jgit.internal.storage.file.GC.writePack(GC.java:733)
        at org.eclipse.jgit.internal.storage.file.GC.repack(GC.java:577)
        at org.eclipse.jgit.internal.storage.file.GC.gc(GC.java:170)
        at org.eclipse.jgit.api.GarbageCollectCommand.call(GarbageCollectCommand
.java:175)
        ... 14 more

**Comment by javabrett** [comment]
Duplicates #24, #38 and #71.  This is caused by a JGit bug, fixed in 4.4, where it cares about what is in `ORIG_HEAD` when running JGit `git gc`, whereas native `git gc` does not.  You have rebased at some point and your `ORIG_HEAD` points to something that is no longer in the repo.

There's a BFG code-workaround as #146, but certainly easier for you to just remove/reset your `ORIG_HEAD` to `HEAD` manually, then run BFG again and it won't crash.

It looks like JGit 4.4 is released, so BFG should probably consume that now.

**Comment by rtyley** [comment]
> It looks like JGit 4.4 is released, so BFG should probably consume that now.

I had a quick look at that this morning - unfortunately, it appears that somewhere between JGit `4.2.0.201601211800-r` and `4.3.0.201604071810-r` there was a change that now causes two tests in the BFG test suite to fail:

```
[info] x remove all big blobs, even if they have identical size
[error]    Set(AnyObjectId[06d7405020018ddf3cacee90fd4af10487da3d20], AnyObjectId[cb2cd7aa871a4a9d56bfb3495b1dc817b5b7589d]) is not empty (unpackedRepo.scala:80)
[info] 
[info] x convert big blobs to the Git LFS format
[error]    Set(AnyObjectId[596cfc4d9e27d0a77adf744a7dbdc76bab37b4a9]) is not empty (unpackedRepo.scala:80)
```

**Comment by javabrett** [comment]
Heh yes I had noticed the same earlier today and c can confirm. Need to find out if they have a gc() regression along with the fix.

**Comment by javabrett** [comment]
For the failed tests, if you remove the [first `gc()` call in `ensureRemovalOfBadEggs`](https://github.com/rtyley/bfg-repo-cleaner/blob/master/bfg-test/src/main/scala/com/madgag/git/bfg/test/unpackedRepo.scala#L74), the tests pass.  Need to establish what goes wrong in newer `gc()`, although that `gc()` call is superfluous.

**Comment by javabrett** [comment]
The behavioural difference is that with the newest JGit, after the test fails, the repo has two packs instead of one:

```
./objects/pack/pack-f1a94bd7e851d20777565dc8ed77a6efb5549680.bitmap
./objects/pack/pack-3ca22ee16a80f9dff4c17bd2b269741eadc277d1.idx
./objects/pack/pack-f1a94bd7e851d20777565dc8ed77a6efb5549680.idx
./objects/pack/pack-3ca22ee16a80f9dff4c17bd2b269741eadc277d1.pack
./objects/pack/pack-f1a94bd7e851d20777565dc8ed77a6efb5549680.pack
./objects/pack/pack-3ca22ee16a80f9dff4c17bd2b269741eadc277d1.bitmap
```

`f1a94bd7e851d20777565dc8ed77a6efb5549680` is the original pack (same after first `gc()`, `3ca22ee16a80f9dff4c17bd2b269741eadc277d1` is new.

I expected the `gc()` to be blocking.  I can't reproduce this using JGit's CLI tool after a BFG run, with either 4.2.0 or 4.4.0.

**Comment by javabrett** [comment]
Created #160 to track JGit upgrade.

**Comment by javabrett** [comment]
I worked-out what the test-failure is caused by, and how to fix it.  I will write it up, and it is probably worthy of some discussion on the JGit lists.

It relates to changes that were made to make pack-removal during GC incorporate an expiry-time test, and then a later config change for that feature.  This results in BFG not getting correct-removal of the packs containing old, rewritten history, causing the tests to fail when those packs are retained.  This is further complicated by the pack-dates being taken from disk-stats, which are fresh because they were just unpacked then `gc()`ed.

**Comment by javabrett** [comment]
Raised https://bugs.eclipse.org/bugs/show_bug.cgi?id=497227 for the pack-expiry issue.

**Comment by rtyley** [comment]
Thanks for all your work on this @javabrett - I've just released BFG v1.12.13 with JGit 4.4.1, so hopefully this issue should be resolved...

**Comment by javabrett** [comment]
AFAIK this issue is resolved and can be closed.
